src/MonoTorrent/MonoTorrent.Streaming/LocalStream.cs

```diff
@@ -0,0 +1,155 @@
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+using MonoTorrent.Client;
+using MonoTorrent.Client.PiecePicking;
+
+namespace MonoTorrent.Streaming
+{
+    /// <summary>
+    /// A seekable Stream which can be used to access a <see cref="TorrentFile"/> while it is downloading.
+    /// If the stream seeks to a location which hasn't been downloaded yet, <see cref="Read(byte[], int, int)"/>
+    /// will block until the data is available. <see cref="ReadAsync(byte[], int, int, CancellationToken)"/>
+    /// will perform a non-blocking wait for the data.
+    /// </summary>
+    class LocalStream : Stream
+    {
+        long position;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length => File.Length;
+
+        public override long Position {
+            get => position;
+            set => Seek (value, SeekOrigin.Begin);
+        }
+
+        TorrentFile File { get; }
+
+        TorrentManager Manager { get; }
+
+        StreamingPiecePicker Picker { get; }
+
+        FileStream Stream { get; set; }
+
+        public LocalStream (TorrentManager manager, TorrentFile file, StreamingPiecePicker picker)
+        {
+            Manager = manager;
+            File = file;
+            Picker = picker;
+        }
+
+        protected override void Dispose (bool disposing)
+        {
+            base.Dispose (disposing);
+            if (disposing) {
+                Stream?.Dispose ();
+            }
+        }
+
+        public override void Flush ()
+            => throw new NotSupportedException ();
+
+        public override int Read (byte[] buffer, int offset, int count)
+            => ReadAsync (buffer, offset, count, CancellationToken.None).GetAwaiter ().GetResult ();
+
+        public override async Task<int> ReadAsync (byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // The torrent is treated as one big block of data, so this is the offset at which the current file's data starts at.
+            var torrentFileStartOffset = (long)File.StartPieceIndex * (long)Manager.Torrent.PieceLength + File.StartPieceOffset;
+
+            // Take our current position into account when calculating the start/end pieces of the data we're reading.
+            var startPiece = (int) (torrentFileStartOffset + Position) / Manager.Torrent.PieceLength;
+            var endPiece = (int) (torrentFileStartOffset + Position + count) / Manager.Torrent.PieceLength;
+            while (Manager.State != TorrentState.Stopped && Manager.State != TorrentState.Error) {
+                bool allAvailable = true;
+                for (int i = startPiece; i <= endPiece && allAvailable; i++)
```

Comment:
I guess there are other places with similar code. Maybe it should be moved to something like `bool BitField.AllInRange(int startPiece, int count)`? Name like `AnyInRange` for OR can be used too.